homeassistant/components/onvif/camera.py

```diff
@@ -307,37 +356,73 @@ def __init__(self, hass, config):
     def setup_ptz(self):
         """Set up PTZ if available."""
         _LOGGER.debug("Setting up the ONVIF PTZ service")
-        if self._camera.get_service("ptz", create=False) is None:
+        if self._camera.get_service("ptz") is None:
             _LOGGER.debug("PTZ is not available")
         else:
             self._ptz_service = self._camera.create_ptz_service()
-            _LOGGER.debug("Completed set up of the ONVIF camera component")
+        _LOGGER.debug("Completed set up of the ONVIF camera component")
 
-    async def async_perform_ptz(self, pan, tilt, zoom):
+    async def async_perform_ptz(
+        self, pan, tilt, zoom, distance, speed, move_mode, continuous_duration
+    ):
         """Perform a PTZ action on the camera."""
         if self._ptz_service is None:
             _LOGGER.warning("PTZ actions are not supported on camera '%s'", self._name)
             return
 
         if self._ptz_service:
-            pan_val = 1 if pan == DIR_RIGHT else -1 if pan == DIR_LEFT else 0
-            tilt_val = 1 if tilt == DIR_UP else -1 if tilt == DIR_DOWN else 0
-            zoom_val = 1 if zoom == ZOOM_IN else -1 if zoom == ZOOM_OUT else 0
-            req = {
-                "Velocity": {
-                    "PanTilt": {"_x": pan_val, "_y": tilt_val},
-                    "Zoom": {"_x": zoom_val},
-                }
-            }
+            pan_val = (
+                distance if pan == DIR_RIGHT else -distance if pan == DIR_LEFT else 0
```

Comment:
Instead of this long if block we can make a map between the pan direction and the factor to multiply the distance.

```py
PAN_FACTOR = {
    DIR_RIGHT: 1,
    DIR_LEFT: - 1,
}

...

pan_val = distance * PAN_FACTOR.get(pan, 0)
```